Please note, the go linter is failing with the following error:

```
 Error: pkg/ipam/service/ipallocator/allocator_test.go:15:42: non-wrapping format verb for fmt.Errorf. Use `%w` to format errors (errorlint)
  		panic(fmt.Errorf("net.ParseCIDR: %+v", err))
```

--- a/Documentation/network/concepts/ipam/multi-pool.rst
+++ b/Documentation/network/concepts/ipam/multi-pool.rst
@@ -71,7 +71,10 @@ New pools can be added at run-time. The list of CIDRs in each pool can also be
 extended at run-time. In-use CIDRs may not be removed from an existing pool, and
 existing pools may not be deleted if they are still in use by a Cilium node.
 The mask size of a pool is immutable and the same for all nodes. Neither restriction
-is enforced until :gh-issue:`26966` is resolved.
+is enforced until :gh-issue:`26966` is resolved. The first and last address of a
+``CiliumPodIPPool`` are reserved and cannot be allocated. Pools with less than 3
+addresses (/31, /32, /127, /128) do not have this limitation.
+
 
 Configuration
 *************

--- a/pkg/ipam/allocator/multipool/pool_allocator.go
+++ b/pkg/ipam/allocator/multipool/pool_allocator.go
@@ -67,13 +67,25 @@ func (m errAllocatorNotReady) Is(target error) bool {
 	return errors.Is(target, ErrAllocatorNotReady)
 }
 
+// addrsInPrefix calculates the number of usable addresses in a prefix p, or 0 if p is not valid.
 func addrsInPrefix(p netip.Prefix) *big.Int {
+	if !p.IsValid() {
+		return big.NewInt(0)
+	}
+
 	// compute number of addresses in prefix, i.e. 2^bits
 	addrs := new(big.Int)
 	addrs.Lsh(big.NewInt(1), uint(p.Addr().BitLen()-p.Bits()))
+
+	// prefix has less than 3 addresses
+	two := big.NewInt(2)
+	if addrs.Cmp(two) <= 0 {
+		return addrs
+	}
+
 	// subtract network and broadcast address, which are not available for
 	// allocation in the cilium/ipam library for now
-	addrs.Sub(addrs, big.NewInt(2))
+	addrs.Sub(addrs, two)
 	if addrs.Sign() < 0 {
 		return big.NewInt(0)
 	}

--- a/pkg/ipam/allocator/multipool/pool_allocator_test.go
+++ b/pkg/ipam/allocator/multipool/pool_allocator_test.go
@@ -476,19 +476,19 @@ func Test_addrsInPrefix(t *testing.T) {
 			want: big.NewInt(0),
 		},
 		{
-			name: "two",
-			args: netip.MustParsePrefix("10.0.0.0/30"),
+			name: "/32",
+			args: netip.MustParsePrefix("10.0.0.0/32"),
+			want: big.NewInt(1),
+		},
+		{
+			name: "/31",
+			args: netip.MustParsePrefix("10.0.0.0/31"),
 			want: big.NewInt(2),
 		},
 		{
-			name: "underflow /31",
-			args: netip.MustParsePrefix("10.0.0.0/31"),
-			want: big.NewInt(0),
-		},
-		{
-			name: "underflow /32",
-			args: netip.MustParsePrefix("10.0.0.0/32"),
-			want: big.NewInt(0),
+			name: "/30",
+			args: netip.MustParsePrefix("10.0.0.0/30"),
+			want: big.NewInt(2),
 		},
 	}
 	for _, tt := range tests {

--- a/pkg/ipam/service/ipallocator/allocator.go
+++ b/pkg/ipam/service/ipallocator/allocator.go
@@ -65,24 +65,26 @@ type Range struct {
 }
 
 // NewCIDRRange creates a Range over a net.IPNet, calling allocator.NewAllocationMap to construct
-// the backing store.
+// the backing store. Returned Range excludes first (base) and last addresses (max) if provided cidr
+// has more than 2 addresses.
 func NewCIDRRange(cidr *net.IPNet) *Range {
 	base := bigForIP(cidr.IP)
-	max := maximum(0, int(RangeSize(cidr)-2)) // don't use the network broadcast,
+	size := RangeSize(cidr)
+
+	// for any CIDR other than /32 or /128:
+	if size > 2 {
+		// don't use the network broadcast
+		size = max(0, size-2)
+		// don't use the network base
+		base = base.Add(base, big.NewInt(1))
+	}
 
 	return &Range{
 		net:   cidr,
-		base:  base.Add(base, big.NewInt(1)), // don't use the network base
-		max:   max,
-		alloc: allocator.NewAllocationMap(int(max), cidr.String()),
+		base:  base,
+		max:   int(size),
+		alloc: allocator.NewAllocationMap(int(size), cidr.String()),
 	}
-}
-
-func maximum(a, b int) int {
-	if a > b {
-		return a
-	}
-	return b
 }
 
 // Free returns the count of IP addresses left in the range.

--- a/pkg/ipam/service/ipallocator/allocator_test.go
+++ b/pkg/ipam/service/ipallocator/allocator_test.go
@@ -1,0 +1,122 @@
+package ipallocator
+
+import (
+	"fmt"
+	"math/big"
+	"net"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func mustParseCidr(cidr string) *net.IPNet {
+	_, ipNet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		panic(fmt.Errorf("net.ParseCIDR: %w", err))
+	}
+	return ipNet
+}
+
+func ipForBig(i *big.Int) net.IP {
+	return addIPOffset(i, 0)
+}
+
+func TestNewCIDRRange(t *testing.T) {
+	testCases := []struct {
+		name     string
+		ipNet    *net.IPNet
+		wantBase net.IP
+		wantMax  int
+	}{
+		{
+			name:     "IPv4 /27",
+			ipNet:    mustParseCidr("192.168.0.1/27"),
+			wantBase: net.ParseIP("192.168.0.1"),
+			wantMax:  30, // (2^(32-27) - 2
+		},
+		{
+			name:     "IPv4 /31",
+			ipNet:    mustParseCidr("192.168.0.1/31"),
+			wantBase: net.ParseIP("192.168.0.0"),
+			wantMax:  2, // 2^1
+		},
+		{
+			name:     "IPv4 /32",
+			ipNet:    mustParseCidr("192.168.0.1/32"),
+			wantBase: net.ParseIP("192.168.0.1"),
+			wantMax:  1, // 2^0
+		},
+		{
+			name:     "IPv6 /64",
+			ipNet:    mustParseCidr("2001:db8::1/64"),
+			wantBase: net.ParseIP("2001:db8::1"),
+			wantMax:  65534, // max(2^(128-64), 65536) - 2
+		},
+		{
+			name:     "IPv6 /120",
+			ipNet:    mustParseCidr("2001:db8::1/120"),
+			wantBase: net.ParseIP("2001:db8::1"),
+			wantMax:  254, // 2^(128-120) - 2
+		},
+		{
+			name:     "IPv6 /127",
+			ipNet:    mustParseCidr("2001:db8::1/127"),
+			wantBase: net.ParseIP("2001:db8::0"),
+			wantMax:  2, // 2^1
+		},
+		{
+			name:     "IPv6 /128",
+			ipNet:    mustParseCidr("2001:db8::1/128"),
+			wantBase: net.ParseIP("2001:db8::1"),
+			wantMax:  1, // 2^0
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual := NewCIDRRange(tc.ipNet)
+			baseIP := ipForBig(actual.base)
+			require.Equal(t, tc.wantBase.String(), baseIP.String())
+			require.Equal(t, tc.wantMax, actual.max)
+		})
+	}
+}
+
+func TestRangeSize(t *testing.T) {
+	testCases := []struct {
+		name  string
+		ipNet *net.IPNet
+		want  int64
+	}{
+		{
+			name:  "IPv4 /27",
+			ipNet: mustParseCidr("192.168.0.0/27"),
+			want:  32,
+		},
+		{
+			name:  "IPv4 /32",
+			ipNet: mustParseCidr("192.168.0.0/32"),
+			want:  1,
+		},
+		{
+			name:  "IPv6 /64",
+			ipNet: mustParseCidr("2001:db8::/64"),
+			want:  65536,
+		},
+		{
+			name:  "IPv6 /120",
+			ipNet: mustParseCidr("2001:db8::/120"),
+			want:  256,
+		},
+		{
+			name:  "IPv6 /128",
+			ipNet: mustParseCidr("2001:db8::/128"),
+			want:  1,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actual := RangeSize(tc.ipNet)
+			require.Equal(t, tc.want, actual)
+		})
+	}
+}

